File: cs/Minsk/CodeAnalysis/Syntax/SyntaxToken.cs
using System.Text;

namespace Minsk.CodeAnalysis.Syntax;

public sealed class SyntaxToken : SyntaxNode
{
    public SyntaxToken(SyntaxKind kind, string text, int position, object? value)
    {
        Kind = kind;
        Text = text;
        Position = position;
        Value = value;
    }

    public override SyntaxKind Kind { get; }
    protected override IEnumerable<SyntaxNode> Children => Enumerable.Empty<SyntaxNode>();
    public string Text { get; }
    public int Position { get; }
    public object? Value { get; }
    public TextSpan Span => new(Position, Text.Length);

    public override string ToString()
    {
        var result = new StringBuilder();
        result.Append($"{Kind} '{Text}'");
        if (Value != null)
        {
            result.Append($" {Value}");
        }

        return result.ToString();
    }
}

File: nim/src/minsk/codeAnalysis/syntax/syntaxToken.nim
import minsk/codeAnalysis/textSpan
import minsk/minskObject

import syntaxKind
import syntaxNode

type
  SyntaxToken* = ref object of SyntaxNode
    mKind: SyntaxKind
    position*: int
    text*: string
    value*: MinskObject

proc newToken*(
  kind: SyntaxKind,
  position: int,
  text: string,
  value: MinskObject
): SyntaxToken =
  new(result)
  result.mKind = kind
  result.position = position
  result.text = text
  result.value = value

proc span*(self: SyntaxToken): TextSpan =
  newTextSpan(self.position, self.text.len)

method kind*(self: SyntaxToken): SyntaxKind =
  self.mKind

method children*(self: SyntaxToken): seq[SyntaxNode] =
  @[]
